0003-secret-seed-storage-and-key-derivation.md

```diff
@@ -0,0 +1,195 @@
+This system was invented ad-hoc and the motivation for this spike is to design a well thought out solution that is informed by other projects.
+
+I present the conclusions first and a summary of what I learned from LND.
+
+## Secret Seed Storage
+
+This section recommends how we should generate and store the secret seed by answering key questions.
+
+### Is there any way to recover anything if you have your secret seed but no database?
+
+**In practice, No.**
+
+In Theory and in the case of Ethereum only you could scan the blockchain for any HTLC looking contract paying money to an address you generated deterministically from your seed and reclaim it.
+
+In Bitcoin you can't because you'll never be able to compute the P2WSH scriptPubKey without knowing the other person's input.
+If it has already been redeemed, you could scan for redeem transactions with script that matches our Bitcoin HTLC, recover the secret and then look for the HTLC on the other chain (as above, assuming it's Ethereum).
+
+I don't think this partial recovery is worth putting the significant effort that it would require to implement.
+
+### Should it be encrypted?
+
+I think this should be optional.
+A COMIT node doesn't own the funds long term.
+It can only control the funds of protocols that are in being executed.
+This is different to LND where if someone gets access to your seed they can take all your funds regardless of whether you have channels open or not.
+
+The best attack vector against an unencrypted seed is for the attacker to get short term access to the system COMIT is running on, covertly steal their secret seed and then wait for them to execute a swap and then redeem to your own address.
+Encrypting the seed does give you protection against this.
+
+Having said that, giving an attacker access to your COMIT node will probably give them access to the HTTP API which they can use to redeem funds to their own address.
+This is probably a better attack vector than the above.
+
+note: We could prevent this attack by giving the node a way of generating the final redeem addresses without getting them from the api.
+
+I think we should leave it unencrypted for now, make encryption an option later on and then perhaps make encryption the default as a last step.
+
+### Where to store it?
+
+Unlike lnd, I don't think we should store secret seed in the database file.
+This allows users to backup the secret seed separately from the database (although they need to backup both).
+Given that the two are linked, we should store a hash of the secret seed along with the swap id (in the same row or whatever).
+This allows the user to change the secret seed while not invalidating the entire database.
+
+By default I'd put it somewhere like:
+
+```
+~/.comit/secret_seed
+```
+
+But its location should be configurable with `secret_seed_file` in the `.toml` config.
+
+The format of the file should be toml with a single entry `secret_seed` (so we can `encrypted_secret_seed` later).
```

Comment:
What do you think of allowing to pass in an encrypted file itself instead?

Reply:
I am also in favor of just passing in an encrypted file that contains the seed in binary format. No need for a toml IMO.

Reply:
I'm not sure I understood the suggestions above. What about if it's not encrypted? Do you mean in the main config file you you `secret_seed_file = /path/to/file`  OR `encrypted_secret_seed_file = /path/to/encrypted_file`.

Although this would work, I think I'd prefer having it in a toml instead because then you can transition from encrypted to unencrypted without having to edit your main config file.

Reply:
Instead of having two toml entries, I'd rather design the seed-file itself in a way so that you can tell wether or not it is encrypted.

One way would be file-extension. A header in the file itself would also work.
In any case, that seems to be a much smoother way for the user to transition, since they don't have to change the config at all and there is always just one entry that refers to the `seed_file`.

Reply:
I'd prefer not to invent a new file format and just use something that exists.
Toml certainly might not be the best format!

Reply:
Can someone explain what cons they see for a Toml file?

Reply:
> I'd prefer not to invent a new file format and just use something that exists.

I am pretty sure there are standard we can use for this.

> Can someone explain what cons they see for a Toml file?

Toml is a format for storing structured data. If all we have is a single entry, such a file format seems ill-fitted and overly complicated.

Reply:
@thomaseizinger The thing is we have at least two entries: `secret_seed` and `encrypted_secret_seed`. We could use [pem](https://docs.rs/pem/0.6.0/pem/) to label one or the other.

Reply:
I changed it to say the format should be text like the pem format.